beluga_ros/include/beluga_ros/tf2_sophus.hpp

```diff
@@ -245,6 +269,28 @@ inline Sophus::Matrix3<Scalar>& covarianceRowMajorToEigen(const Array<Scalar, 36
   return out;
 }
 
+/// Converts an SE3 `pose` and its covariance to a ROS message.
+/// NOTE: Input covariance is expected to use tangent space parametrization, consistent with the one in beluga's \c
+/// estimation libraries.
+template <typename Scalar>
+inline beluga_ros::msg::PoseWithCovariance toMsg(
+    const Sophus::SE3<Scalar>& in,
+    const Eigen::Matrix<Scalar, 6, 6>& covariance) {
+  beluga_ros::msg::PoseWithCovariance out;
+  tf2::toMsg(in, out.pose);
+
+  // ROS covariances use extrinsic RPY parametrization, we need to convert to it from our tangent space
+  // representation. We use an unscented transform to apply such transform as it's non linear.
+  // See https://www.ros.org/reps/rep-0103.html#covariance-representation .
+
+  const auto& [base_pose_in_map_xyz_rpy, base_pose_covariance_xyz_rpy] =
+      beluga::unscented_transform(in.log(), covariance, detail::TangentialSpaceToXyzRpy<Scalar>);
+  std::copy(
+      base_pose_covariance_xyz_rpy.data(), base_pose_covariance_xyz_rpy.data() + base_pose_covariance_xyz_rpy.size(),
+      out.covariance.data());
```

Comment:
@serraramiro1 meta: are we sure storage layout matches between Eigen and ROS messages? I vaguely recall Eigen using column major layouts by default, and ROS messages using row major layouts.

Reply:
You're totally right. I was 100% sure Eigen was row major too, turns out I was wrong.

Fixed!